### PR TITLE
Cloud registry credentials support

### PR DESCRIPTION
**Pull request by sonnysideup**
## Design

- Implements a generic means of fetching cloud registry credentials (e.g. ECR, GCR, ACR) will support for ECR as a first run
- Controller processes new `DynamicCloudCredentials` field on CR and uses that to fetch credentials and add them to the build job as a new mounted secret
- Build job checks if credentials exist in a known location and adds them to _dynamic_ registries
- Username/Password processing details (from auth token et al) is handled by a separate func to create a standard interface for build job

### CODE REVIEW

File: internal/credentials/docker.go [hunk 1]
```diff
@@ -1,13 +1,56 @@
...
-// DockerConfigJSON models the structure of .dockerconfigfile data
+// DockerConfigJSON models the structure of .dockerconfigfile data.
 type DockerConfigJSON struct {
-	Auths DockerConfig `json:"auths"`
+	Auths AuthConfigs `json:"auths"`
+}
+
+// ExtractDockerAuth will return a username/password combination from a JSON-representation of a Docker config file.
+// This function will process credentials from any of the possible auth-sources within that configuration.
+func ExtractDockerAuth(input []byte, host string) (string, string, error) {
```
Comment (steved) [inline internal/credentials/docker.go:23]: Can we use `docker/cli` for this or does the API not quite fit?

File: internal/cloud/registry/acr/acr.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+	"github.com/docker/docker/api/types"
+)
+
+// AuthN details: https://docs.microsoft.com/en-us/azure/container-registry/container-registry-authentication
+
+//var urlRegex = regexp.MustCompile(`^acr\.io$`)
+
+func LoadAuths(ctx context.Context, url string) (*types.AuthConfig, error) {
+	return nil, errors.New("GCR is unsupported")
```
Comment (steved) [inline internal/cloud/registry/acr/acr.go:15]: nitpick, ACR

File: internal/cloud/registry/ecr/ecr.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/aws/aws-sdk-go-v2/service/ecr"
+	"github.com/aws/aws-sdk-go-v2/service/ecr/ecriface"
+	"github.com/docker/docker/api/types"
+	"github.com/pkg/errors"
+
+	"github.com/dominodatalab/forge/internal/cloud/registry"
+)
+
+var (
+	urlRegex = regexp.MustCompile(`^\d{12}\.dkr\.ecr\.[a-z0-9-]+\.amazonaws.com$`)
```
Comment (steved) [inline internal/cloud/registry/ecr/ecr.go:18]: https://github.com/awslabs/amazon-ecr-credential-helper/blob/10153a961e2dd9da9f40f6cc284572252b199c58/ecr-login/api/client.go#L34 is a bit more complicated. I wonder if in general the best way to find the "correct" loader is to just try and grab a token.

File: internal/cloud/registry/gcr/gcr.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	//	Username: "oauth2accesstoken",
+	//	Password: token.AccessToken,
+	//}
+
+	return nil, nil
+}
+
+//func init() {
+//	registry.DefaultURLMux().RegisterLoader(urlRegex, LoadAuths)
+//}
```
Comment (steved) [inline internal/cloud/registry/gcr/gcr.go:45]: Maybe pull these out into a separate PR for now? Just doesn't seem right to merge the GCR & ACR files.

**Review by steved — APPROVED** [review]
Minor questions, LGTM overall!

File: internal/cloud/registry/ecr/ecr.go (hunk 3 above)
Comment (sonnysideup) [reply internal/cloud/registry/ecr/ecr.go:18]: It seems like using the regex pattern from the ECR helper is a better option than the one I wrote. That should be an improvement and easy to write tests against that. 

>I wonder if in general the best way to find the "correct" loader is to just try and grab a token.

I guess I don't understand what you mean when you say "just try". Are you suggesting we don't use regex but instead try and perform auth using all loaders until we hit gold? If so, considering the following: (a) blind auth calls can be expensive (e.g. AWS config loading will hang for several secs if there are no valid credentials, not sure about the others) and (b) I'm not sure what the errors returned by the other libraries are (i.e. how do we discern between an error that indicates "not supported" and "supported but failed to auth"?).

Comment (steved) [reply internal/cloud/registry/ecr/ecr.go:18]: Yeah that was the gist. I hadn't considered the performance cost and the potential error handling constraints, thanks.

File: internal/credentials/docker.go (hunk 1 above)
Comment (steved) [reply internal/credentials/docker.go:23]: I was thinking something like (haven't tested it)
```
auth := configfile.New().LoadFromReader(input).GetAuthConfig(host)
return auth.Username, auth.Password, ...
```
It would basically handle the decode / `Auth` stuff for you, not the credential helper stuff.
